### PR TITLE
Create policy consent during registration flow

### PR DESCRIPTION
**Pull request by hwupathum**
### Proposed changes in this pull request
This pull request enhances the user provisioning process to support user consent management during registration, including both accepted and rejected consents. It introduces new methods to handle consent creation, ensures PII categories are created as needed, and updates dependencies to newer versions for consent management and identity framework.

**User Consent Management Enhancements:**

* Added logic to process and persist user consents (both accepted and rejected) during user registration by introducing `createConsent` and `createRejectedConsents` methods in `UserProvisioningExecutor`. These methods interact with the consent management service to record consents and rejected consents for each user. [[1]](diffhunk://#diff-64368beba049a0e47408569fe425113401b0b4453eeed746d5c656970cc37b41R202-R205) [[2]](diffhunk://#diff-64368beba049a0e47408569fe425113401b0b4453eeed746d5c656970cc37b41R354-R459)
* Added helper methods to ensure required PII categories exist when creating consents, creating them on-demand if missing.

**Dependency Updates:**

* Updated the `carbon.identity.framework.version` to `7.10.78` in `pom.xml` for improved compatibility and features.
* Updated the `carbon.consent.mgt.version` to `2.9.1-SNAPSHOT` in `pom.xml` to use the latest consent management features and bug fixes.

**Constants and Imports:**

* Added new constants and imports to support consent management, including consent states, prefixes, and error codes. [[1]](diffhunk://#diff-64368beba049a0e47408569fe425113401b0b4453eeed746d5c656970cc37b41R25-R27) [[2]](diffhunk://#diff-64368beba049a0e47408569fe425113401b0b4453eeed746d5c656970cc37b41R66-R78) [[3]](diffhunk://#diff-64368beba049a0e47408569fe425113401b0b4453eeed746d5c656970cc37b41R95-R96) [[4]](diffhunk://#diff-64368beba049a0e47408569fe425113401b0b4453eeed746d5c656970cc37b41R109-R110)

## Related issue
- https://github.com/wso2/product-is/issues/26859

## Summary by CodeRabbit

* **New Features**
  * Consent v2 support during registration: per-purpose accepted and rejected consent receipts are created and recorded, with validation and diagnostic events; receipt creation is skipped when required purpose/category bindings are missing.

* **Bug Fixes**
  * Improved privacy by masking usernames in error and exception messages.

* **Chores**
  * Bumped framework and consent-management library versions.

* **Tests**
  * Added unit tests covering Consent v2 processing, error paths, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/wso2-extensions/identity-governance/pull/1122?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java
Comment (hwupathum) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Update here too

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 1]
```diff
@@ -96,8 +110,10 @@
 public class UserProvisioningExecutor implements Executor {
 
     private static final Log LOG = LogFactory.getLog(UserProvisioningExecutor.class);
+    private static final String COMPONENT_ID = "UserProvisioningExecutor";
     private static final String WSO2_CLAIM_DIALECT = "http://wso2.org/claims/";
     private static final String USERNAME_PATTERN_VALIDATION_SKIPPED = "isUsernamePatternValidationSkipped";
+    private static final String SYSTEM_APP_ID = "Resident IDP";
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Use from framework

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 2]
```diff
@@ -189,6 +205,23 @@ private ExecutorResponse handleRegistrationFlow(ExecutorResponse response, FlowE
             String userid = ((AbstractUserStoreManager) userStoreManager).getUserIDFromUserName(user.getUsername());
             user.setUserStoreDomain(userStoreDomainName);
             user.setUserId(userid);
+
+            if (FrameworkUtils.isConsentV2APIEnabled()) {
```
Comment (hwupathum) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: refctror to a pvt mtd

Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Fixed in https://github.com/wso2-extensions/identity-governance/pull/1122/commits/27f02a6a59e5c086bfd54d700e77c049f912e358

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 1 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Fixed in https://github.com/wso2-extensions/identity-governance/pull/1122/commits/27f02a6a59e5c086bfd54d700e77c049f912e358

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 3]
```diff
@@ -356,6 +373,262 @@ private boolean isExistingUser(FlowUser user, FlowExecutionContext context) thro
...
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Default PII category not found for consent type: " + purposeType +
+                            ". Skipping rejected consent processing for user: " + Utils.maskIfRequired(subjectId));
+                }
+                return;
+            }
+            List<PurposePIICategoryBinding> purposeBindings = new ArrayList<>();
+            for (String purposeUuid : rejectedConsents) {
+                if (StringUtils.isBlank(purposeUuid)) {
+                    continue;
```
Comment (Thisara-Welmilla) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: What are the possible scenarios where `purposeUuid` can be blank. In such cases, does ignore is the expected way of handling.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 4]
```diff
@@ -356,6 +373,262 @@ private boolean isExistingUser(FlowUser user, FlowExecutionContext context) thro
...
+    private void processPolicyConsent(String subjectId, String tenantDomain, String purposeUuid, String consentType)
+            throws ConsentManagementException {
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Processing policy consent. Type: " + consentType + ", User: " +
+                    Utils.maskIfRequired(subjectId) + ", Purpose: " + purposeUuid +
+                    ", Tenant: " + tenantDomain);
+        }
+
+        PIICategory piiCategory = ConsentReceiptUtils.getDefaultPiiCategory(consentType,
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Do we need a null check for the piiCategory here?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 5]
```diff
@@ -356,6 +373,262 @@ private boolean isExistingUser(FlowUser user, FlowExecutionContext context) thro
...
+                     null, false, null, null, RESIDENT_IDP, purposeBindings,
+                     IdentityRecoveryServiceDataHolder.getInstance().getConsentManager());
+            try {
+                PrivilegedCarbonContext.startTenantFlow();
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+                PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(subjectId);
+                IdentityRecoveryServiceDataHolder.getInstance().getConsentManager().addConsent(receiptInput);
+            } finally {
+                PrivilegedCarbonContext.endTenantFlow();
+            }
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: This block will each time the loop is iterated in `createConsent`. Please improve this, if possible

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 6]
```diff
@@ -356,6 +373,262 @@ private boolean isExistingUser(FlowUser user, FlowExecutionContext context) thro
...
+                                LoggerUtils.getMaskedContent(username) : username)
+                        .inputParam(LogConstants.InputKeys.TENANT_DOMAIN, tenantDomain)
+                        .inputParam("consent_type", purposeType)
+                        .inputParam(LogConstants.InputKeys.ERROR_MESSAGE, e.getMessage())
+                        .resultMessage("Failed to create accepted consents.")
+                        .logDetailLevel(DiagnosticLog.LogDetailLevel.APPLICATION)
+                        .resultStatus(DiagnosticLog.ResultStatus.FAILED));
+            }
+            LOG.error("Error while creating consent of type: " + purposeType + " for user: " +
+                    Utils.maskIfRequired(username), e);
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Is the consent creation atomic? Do we need to handle the rollback?

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 3 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Fixed in https://github.com/wso2-extensions/identity-governance/pull/1122/commits/0ae1a979960a4fd36923d01853ebaf9e7dd8db3d

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 4 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: If the piiCategory is not available, it will be created and returned.

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 6 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Yes it is atomic

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 7]
```diff
@@ -356,6 +375,246 @@ private boolean isExistingUser(FlowUser user, FlowExecutionContext context) thro
...
+
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Creating accepted consents for user: " + Utils.maskIfRequired(username) +
+                    " with consent count: " + acceptedConsents.size());
+        }
+
+        int purposeCount = 0;
+        PrivilegedCarbonContext.startTenantFlow();
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setTenantDomain(tenantDomain, true);
+        PrivilegedCarbonContext.getThreadLocalCarbonContext().setUsername(username);
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: is it possible to move the tenant flow initialization to `processUserConsent` method (preferably after 389 )

Comment (ThaminduR) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: This will remove the duplicated tenant flow init calls from `createConsent` and `createRejectedConsents`

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java [hunk 8]
```diff
@@ -356,6 +375,196 @@ private boolean isExistingUser(FlowUser user, FlowExecutionContext context) thro
...
+                    Utils.maskIfRequired(subjectId));
+        }
+    }
+
+    private void addConsentReceipt(String subjectId, String tenantDomain, String purposeType,
+                                   List<PurposePIICategoryBinding> purposeBindings, boolean isRejected)
+            throws FlowEngineServerException {
+
+        try {
+            // ApplicationId is SYSTEM — policy consent is system-wide, not per-application.
```
Comment (ThaminduR) [inline components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Shall we improve the comment, referring to RESIDENT_IDP

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 7 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Fixed in https://github.com/wso2-extensions/identity-governance/pull/1122/commits/09f031f7f73a15c7cf9f518c3e1f317cc25205f5

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 5 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Fixed in https://github.com/wso2-extensions/identity-governance/pull/1122/commits/abefb2eb6e35627ca5324b8a9c531ac522ee9152

File: components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java (hunk 8 above)
Comment (hwupathum) [reply components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/executor/UserProvisioningExecutor.java]: Fixed in https://github.com/wso2-extensions/identity-governance/pull/1122/commits/0ae1a979960a4fd36923d01853ebaf9e7dd8db3d

**Review by ThaminduR — APPROVED** [review]
